aibolit/patterns/many_primary_ctors/many_primary_ctors.py

```diff
@@ -19,42 +19,56 @@
 # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 # SOFTWARE.
+# The MIT License (MIT)
+#
+# Copyright (c) 2020 Aibolit
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included
+# in all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
 
 from typing import List
-
-import javalang.ast
-import javalang.parse
-import javalang.tree
-
-from aibolit.types_decl import LineNumber
+from aibolit.ast_framework import ASTNodeType, AST
 from aibolit.utils.ast_builder import build_ast
+from aibolit.ast_framework.ast_node import ASTNode
 
 
 class ManyPrimaryCtors(object):
-    def value(self, filename: str):
-        tree = build_ast(filename)
-
-        return self.__traverse_node(tree)
-
-    def __traverse_node(self, tree: javalang.ast.Node) -> List[LineNumber]:
-        lines: List[LineNumber] = list()
-
-        for _, class_declaration in tree.filter(javalang.tree.ClassDeclaration):
-            primary_ctors = list(filter(_is_primary, class_declaration.constructors))
-
-            if len(primary_ctors) > 1:
-                lines.extend(ctor.position.line for ctor in primary_ctors)
+    '''
+    If there is more than one primary
+    constructors in a class, it is
+    considered a pattern
+    '''
+    def value(self, filename: str) -> List[int]:
+        lines: List[int] = list()
+        ast = AST.build_from_javalang(build_ast(filename))
+        for class_declaration in ast.get_proxy_nodes(ASTNodeType.CLASS_DECLARATION):
+            primary_lines = self.__find_primary(ast, class_declaration.body)
+            if len(primary_lines) > 1:
+                lines.extend(primary_lines)
 
         return lines
 
+    def __find_primary(self, ast: AST, class_body: List[ASTNode]) -> List[int]:
+        lines: List[int] = []
+        for node in class_body:
+            if node.node_type == ASTNodeType.CONSTRUCTOR_DECLARATION:
+                for assignment in ast.get_subtree(node).get_proxy_nodes(ASTNodeType.ASSIGNMENT):
```

Comment:
According to Yegor post, I think it is easier to detect secondary constructors. They consist only of a single statement. All other constructors are primary.

Reply:
at the moment private constructors are implemented correctly. Constructor is private when we have at least one variable assignment. Secondary constructor is constructor  when we do not have variable assignments and we have a call of primary constructor